src/jenkins.py

```diff
@@ -705,3 +709,195 @@ def get_agent_name(unit_name: str) -> str:
+    )
+
+
+def _traverse_dependencies(
+    plugin: str, dependency_lookup: typing.Mapping[str, typing.Iterable[str]], seen: set[str]
+) -> typing.Iterable[str]:
+    """Recursively traverse through plugin dependencies.
+
+    Args:
+        plugin: The plugin to recurse through the dependencies.
+        dependency_lookup: The plugin and its dependencies lookup table.
+        seen: a set indicating whether a plugin has already been visited.
+
+    Yields:
+        Plugin and its dependents.
+    """
+    if plugin in seen or plugin not in dependency_lookup:
+        return
+    yield plugin
+    seen.add(plugin)
+    for dependency in dependency_lookup[plugin]:
+        yield from _traverse_dependencies(dependency, dependency_lookup, seen)
+
+
+def _get_allowed_plugins(
+    top_level_plugins: typing.Iterable[str],
+    dependency_lookup: typing.Mapping[str, typing.Iterable[str]],
+) -> typing.Iterable[str]:
+    """Get the plugin short names of allowed plugins and their dependencies.
+
+    Args:
+        top_level_plugins: The allowed plugins short names to add to allowed plugins with their
+            dependencies.
+        dependency_lookup: The plugin dependency lookup table.
+
+    Yields:
+        The allowed plugin short name.
+    """
+    seen: set[str] = set()
+    for plugin in top_level_plugins:
+        if plugin not in seen:
+            yield from _traverse_dependencies(plugin, dependency_lookup, seen)
+
+
+def _get_top_level_plugins(
+    plugins: typing.Iterable[str], dependency_lookup: typing.Mapping[str, typing.Iterable[str]]
+) -> typing.Iterable[str]:
+    """Get top level plugins that are not dependencies of other plugins.
+
+    Args:
+        plugins: Plugins to extract top level plugins from.
+        dependency_lookup: The dependency lookup table.
+
+    Returns:
+        All plugins that are not dependency of another plugin.
+    """
+    dependent_plugins: set[str] = set()
+    for _, dependencies in dependency_lookup.items():
+        dependent_plugins = dependent_plugins.union(dependencies)
+
+    return set(plugins) - dependent_plugins
+
+
+def remove_unlisted_plugins(
+    plugins: typing.Iterable[str] | None,
+    container: ops.Container,
+    client: jenkinsapi.jenkins.Jenkins | None = None,
+) -> None:
+    """Remove plugins that are not in the list of desired plugins.
+
+    Args:
+        plugins: The list of plugins that can be installed.
+        container: The workload container.
+        client: The Jenkins API client.
+
+    Raises:
+        JenkinsPluginError: if there was an error removing unlisted plugin.
+        JenkinsError: if there was an error restarting Jenkins after removing the plugin.
+        TimeoutError: if it took too long to restart Jenkins after removing the plugin.
+    """
+    if not plugins:
+        return
+
+    client = client if client is not None else _get_client(get_admin_credentials(container))
```

Comment:
This is a bit weird, when would the client be `None` and not?

Reply:
The client is `None` when the charm calls it, the client is `MagicMock(spec=jenkinsapi.Jenkins)` in unit tests :D, this is only here to facilitate the tests.

Reply:
This will be updated in a follow up refactor PR.